lib/Conversion/StandardToHandshake/StandardToHandshake.cpp

```diff
@@ -631,6 +644,421 @@ int getBranchCount(Value val, Block *block) {
+  // be connected to the exit network. In doing so, we parity-correct these
+  // condition values based on the convention used in buildContinueNetwork - The
+  // loop mux will on select "0" select external control, and internal control
+  // at "1". This convention which must be followed by the loop exit network.
+  // External control must be selected when exiting the loop (to reprime the
+  // register).
+  SmallVector<Value> parityCorrectedConds;
+  for (auto &[condBlock, exitBlock] : exitPairs) {
+    auto condBr = getControlCondBranch(condBlock);
+    assert(
+        condBr &&
+        "Expected a conditional control branch op in the loop condition block");
+    Operation *trueUser = *condBr.trueResult().getUsers().begin();
+    Operation *falseUser = *condBr.falseResult().getUsers().begin();
+    bool isTrueParity = trueUser->getBlock() == exitBlock;
+    bool isFalseParity = falseUser->getBlock() == exitBlock;
+    assert(isTrueParity ^ isFalseParity &&
+           "The user of either the true or the false result should be in the "
+           "exit block");
+
+    Value condValue = condBr.conditionOperand();
+    if (isTrueParity) {
+      // This goes against the convention, and we have to invert the condition
+      // value before connecting it to the exit network.
+      rewriter->setInsertionPoint(condBr);
+      condValue = rewriter->create<arith::XOrIOp>(
+          loc, condValue.getType(), condValue,
+          rewriter->create<arith::ConstantOp>(
+              loc, rewriter->getIntegerAttr(rewriter->getI1Type(), 1)));
+    }
+    parityCorrectedConds.push_back(condValue);
+  }
+
+  // Merge all of the parity-corrected exit conditions and assign them
+  // to the loop priming input.
+  auto exitMerge = rewriter->create<MergeOp>(loc, parityCorrectedConds);
+  loopPrimingInput.setValue(exitMerge);
+}
+
+LogicalResult LoopNetworkRewriter::processOuterLoop(Location loc,
+                                                    Block *loopHeader,
+                                                    Block *loopLatch) {
+  // To build the exit network, we need to determine the exit blocks of the
+  // loop, as well as all blocks contained within the loop. Exit blocks are the
+  // blocks that control is transfered to after exiting the loop. This is
+  // essentially determining the strongly connected components with the loop
+  // header. we perform a BFS from the loop header, and if the loop header is
+  // reachable the block, it is within the loop.
+  SetVector<Block *> inLoop, exitBlocks;
+  blockBFS(loopHeader, [&](Block *currBlock) {
+    if (isReachable(currBlock, loopHeader)) {
+      inLoop.insert(currBlock);
+      return BFSNextState::Continue;
+    }
+    exitBlocks.insert(currBlock);
+    return BFSNextState::SkipSuccessors;
+  });
+
+  assert(inLoop.size() >= 2 && "A loop must have at least 2 blocks");
+  assert(exitBlocks.size() != 0 && "No exit blocks in the loop...?");
```

Comment:
nit: maybe rephrase this like the other assertion? "A loop must have an exit block"